src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,239 @@
+        }
+
+        private static bool  TryGetCodeElementToRename(string oldName,
+                                                       EnvDTE.CodeElements elements,
+                                                       bool isCaseSensitive,
+                                                       out EnvDTE80.CodeElement2 codeElementToRename)
+        {
+            foreach (EnvDTE.CodeElement element in elements)
+            {
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementNamespace)
+                {
+                    if (TryGetCodeElementToRename(oldName, element.Children, isCaseSensitive, out codeElementToRename))
+                    {
+                        return true;
+                    }
+                }
+
+                if (element.Kind == EnvDTE.vsCMElement.vsCMElementStruct ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementClass ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementModule ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementInterface ||
+                    element.Kind == EnvDTE.vsCMElement.vsCMElementEnum)
+                {
+                    string elementName = element.Name;
+                    if(string.Equals(oldName, elementName, isCaseSensitive ? StringComparison.Ordinal: StringComparison.OrdinalIgnoreCase))
+                    {
+                        codeElementToRename = element as EnvDTE80.CodeElement2;
+                        return true;
+                    }
+                }
+            }
+
+            codeElementToRename = null;
+            return false;
+        }
+
+        private static async Task<(bool success, ISymbol symbolToRename)> TryGetSymbolToRenameAsync(string oldName,
+            string newFileName,
+            Project project)
+        {
+            if (project is null)
+                return (false, null);
+
+            Document newDocument = GetDocument(project, newFileName);
+            if (newDocument is null)
+                return (false, null);
+
+            SyntaxNode root = await GetRootNode(newDocument);
+            if (root is null)
+                return (false, null);
+
+            SemanticModel semanticModel = await newDocument.GetSemanticModelAsync();
+            if (semanticModel is null)
+                return (false, null);
+
+            IEnumerable<SyntaxNode> declarations = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(semanticModel, n, oldName, semanticModel.Compilation.IsCaseSensitive));
+            SyntaxNode declaration = declarations.FirstOrDefault();
+            if (declaration is null)
+                return (false, null);
+
+            return (true, semanticModel.GetDeclaredSymbol(declaration));
+
+        }
+
+        private static Document GetDocument(Project project, string filePath)
+            => (from d in project.Documents where StringComparers.Paths.Equals(d.FilePath, filePath) select d).FirstOrDefault();
```

Comment:
You can do solution.GetDocumentIdsWithFilePath() and then if needed filter that to ones from your project. We internally maintain a dictionary so it'll be faster than the O(n) scan here.